qiskit_machine_learning/kernels/base_kernel.py

```diff
@@ -0,0 +1,134 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base kernel"""
+
+from __future__ import annotations
+
+from abc import abstractmethod, ABC
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.circuit.library import ZZFeatureMap
+
+
+class BaseKernel(ABC):
+    r"""
+    Abstract class defining the Kernel interface.
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    .. math::
+
+        K(x, y) = \langle f(x), f(y)\rangle.
+
+    Here K is the kernel function, x, y are n dimensional inputs. f is a map from n-dimension
+    to m-dimension space. :math:`\langle x, y \rangle` denotes the dot product.
+    Usually m is much larger than n.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints x and y and feature
+    map f, all of n dimension. This kernel matrix can then be used in classical machine learning
+    algorithms such as support vector classification, spectral clustering or ridge regression.
+    """
+
+    def __init__(self, feature_map: QuantumCircuit = None, enforce_psd: bool = True) -> None:
+        """
+        Args:
+            feature_map: Parameterized circuit to be used as the feature map. If ``None`` is given,
+                :class:`~qiskit.circuit.library.ZZFeatureMap` is used with two qubits. If there's
+                a mismatch in the number of qubits of the feature map and the number of features
+                in the dataset, then the kernel will try to adjust the feature map to reflect the
+                number of features.
+            enforce_psd: Project to closest positive semidefinite matrix if ``x = y``.
+                Default ``True``.
+        """
+        if feature_map is None:
+            feature_map = ZZFeatureMap(2)
+
+        self._feature_map = feature_map
+        self._enforce_psd = enforce_psd
```

Comment:
Would we want to expose these (feature_map and enforce_psd) as public instance vars that the user can get/set to read/update these?

Reply:
Resetting the feature map means reinitializing a fidelity instance with the corresponding circuit. If number of parameters change, this also needs to be accounted for which is especially non-trivial in the trainable case. We thought this through at some point and came to the conclusion that it might be easier to just create a new kernel instance when you need to change the feature map.
With `enforce_psd` you're right, I think that could be made public without issues.

Reply:
Updating the feature map may be non-trivial, that's right. Even for trivial properties, I'm hesitant to provide setters, immutable object safer and easier to maintain. Added some getters.